File: Medium/Construct Tree from Inorder & Preorder/construct-tree-from-inorder-preorder.cpp
//{ Driver Code Starts
//

#include<bits/stdc++.h>
using namespace std;

struct Node
{
	int data;
	struct Node *left;
	struct Node *right;
	
	Node(int x){
	    data = x;
	    left = NULL;
	    right = NULL;
	}
};


void printPostOrder(Node *root)
{
	if(root==NULL)
		return;
	printPostOrder(root->left);
	printPostOrder(root->right);
	cout<<root->data<<" ";
}

// } Driver Code Ends


class Solution{
    public:
    int preIdx=0;
    Node* constr(int in[],int pre[],int is,int ie){
        if (is > ie) 
            return NULL;
        Node *root = new Node(pre[preIdx]);
        preIdx++;
        int inIdx=0;
        for(int i=is;i<=ie;i++){
            if(in[i]==root->data){
                inIdx=i;
                break;
            }
        }
        root->left = constr(in,pre,is,inIdx-1);
        root->right = constr(in,pre,inIdx+1,ie);
        return root;
    }
    Node* buildTree(int in[],int pre[], int n){
        return constr(in,pre,0,n-1);
    }
};



//{ Driver Code Starts.
int main()
{
	int t;
	cin>>t;
	while(t--)
	{
		int n;
		cin>>n;
		
		int inorder[n], preorder[n];
		for(int i=0; i<n; i++)
			cin>> inorder[i];
		for(int i=0; i<n; i++)
			cin>> preorder[i];
		Solution obj;
		Node *root = obj.buildTree(inorder, preorder, n);
		printPostOrder(root);
		cout<< endl;
	}
}

// } Driver Code Ends

File: Medium/Construct Tree from Inorder & Preorder/construct-tree-from-inorder-preorder.py
class Solution:
    def buildtree(self, inorder, preorder,n):
        if not inorder or not preorder:
            return None

        root_val = preorder[0]
        root = Node(root_val)

        idx = inorder.index(root_val)
        lsize = idx
        rsize = len(inorder) - idx - 1

        root.left = self.buildtree(inorder[:idx], preorder[1:1 + lsize],n)
        root.right = self.buildtree(inorder[idx + 1:], preorder[1 + lsize:],n)

        return root

#{ 
 # Driver Code Starts
#Initial Template for Python 3

class Node:
    def __init__(self,val):
        self.data = val
        self.right = None
        self.left = None

def printPostorder(n):
    if n is None:
        return
    printPostorder(n.left)
    printPostorder(n.right)
    print(n.data, end=' ')

if __name__=="__main__":
    t = int(input())
    for _ in range(t):
        n = int(input())
        inorder = [ int(x) for x in input().split() ]
        preorder = [ int(x) for x in input().split() ]
        
        root = Solution().buildtree(inorder, preorder, n)
        printPostorder(root)
        print()

# } Driver Code Ends